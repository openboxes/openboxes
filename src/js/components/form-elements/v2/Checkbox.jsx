import React from 'react';

import PropTypes from 'prop-types';

import InputWrapper from 'wrappers/InputWrapper';

import './style.scss';

const Checkbox = ({
  title,
  tooltip,
  button,
  disabled,
  errorMessage,
  labelPosition,
  id,
  name,
  noWrapper,
  indeterminate,
  ...fieldProps
}) => {
  const checkbox = (
    <input
      id={id || name}
      name={name}
      type="checkbox"
      disabled={disabled}
      className={`checkbox-v2 ${errorMessage ? 'has-errors' : ''}`}
      {...fieldProps}
      checked={fieldProps.value}
      ref={(checkboxRef) => {
        if (checkboxRef) {
          // eslint-disable-next-line no-param-reassign
          checkboxRef.indeterminate = indeterminate;
        }
      }}
    />
  );

  return (
    noWrapper ? checkbox : (
      <InputWrapper
        title={title}
        tooltip={tooltip}
        button={button}
        errorMessage={errorMessage}
        inputId={id || name}
        labelPosition={labelPosition}
      >
        <div className="form-element-checkbox">
          {checkbox}
        </div>
      </InputWrapper>
    )
  );
};

export default Checkbox;

Checkbox.propTypes = {
  // Message which will be shown on the tooltip above the field
  tooltip: PropTypes.shape({
    id: PropTypes.string.isRequired,
    defaultMessage: PropTypes.string.isRequired,
  }),
  // Title displayed above the field
  title: PropTypes.shape({
    id: PropTypes.string.isRequired,
    defaultMessage: PropTypes.string.isRequired,
  }),
  // Button on the right side above the input
  button: PropTypes.shape({
    id: PropTypes.string.isRequired,
    defaultMessage: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
  }),
  // Indicator whether the field should be disabled
  disabled: PropTypes.bool,
  // If the errorMessage is not empty then the field is bordered
  // and the message is displayed under the input
  errorMessage: PropTypes.string,
  // position of input label
  labelPosition: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
  // html element id
  id: PropTypes.string,
  // html element name
  name: PropTypes.string,
  // Indicator whether we want to remove the wrapper from the checkbox
  // it is needed to remove additional padding
  noWrapper: PropTypes.bool,
  indeterminate: PropTypes.bool,
};

Checkbox.defaultProps = {
  tooltip: null,
  title: null,
  button: null,
  errorMessage: null,
  disabled: false,
  id: undefined,
  name: undefined,
  labelPosition: 'right',
  noWrapper: false,
  indeterminate: false,
};
