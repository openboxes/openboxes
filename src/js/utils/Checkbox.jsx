import React from 'react';

import PropTypes from 'prop-types';

import Translate from 'utils/Translate';

const Checkbox = ({
  value, indeterminate, custom, fieldRef, withLabel, label, defaultMessage, ...props
}) => {
  const onChange = (event) => {
    const { checked } = event.target;

    if (props.onChange) {
      props.onChange(checked);
    }
  };

  if (custom) {
    return (
      <div data-testid="custom-checkbox" className="custom-checkbox">
        <label htmlFor={props.id}>
          <input
            id={props.id}
            type="checkbox"
            ref={fieldRef}
            checked={value}
            {...props}
            onChange={onChange}
          />
        </label>
      </div>
    );
  }

  if (withLabel) {
    return (
      <div data-testid="with-label-checkbox" className="d-flex align-items-center">
        <input
          type="checkbox"
          ref={(elem) => {
            if (elem) {
              // eslint-disable-next-line no-param-reassign
              elem.indeterminate = indeterminate;
            }
            if (fieldRef) {
              fieldRef(elem);
            }
          }}
          checked={value}
          {...props}
          onChange={onChange}
        />
        <label htmlFor={props.id} style={{ margin: '0 0 0 5px' }}>
          <Translate id={label} defaultMessage={defaultMessage || label} />
        </label>
      </div>
    );
  }

  return (
    <input
      data-testid="checkbox"
      type="checkbox"
      ref={(elem) => {
        if (elem) {
          // eslint-disable-next-line no-param-reassign
          elem.indeterminate = indeterminate;
        }
        if (fieldRef) {
          fieldRef(elem);
        }
      }}
      checked={value}
      {...props}
      onChange={onChange}
    />
  );
};

export default Checkbox;

Checkbox.propTypes = {
  onChange: PropTypes.func,
  value: PropTypes.bool,
  indeterminate: PropTypes.bool,
  custom: PropTypes.bool,
  id: PropTypes.string,
  fieldRef: PropTypes.func,
  withLabel: PropTypes.bool,
  label: PropTypes.string,
  defaultMessage: PropTypes.string,
};

Checkbox.defaultProps = {
  onChange: null,
  value: null,
  indeterminate: false,
  custom: false,
  id: '',
  fieldRef: undefined,
  withLabel: false,
  label: '',
  defaultMessage: '',
};
