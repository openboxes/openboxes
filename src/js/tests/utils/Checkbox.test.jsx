import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Checkbox from 'utils/Checkbox';

import '@testing-library/jest-dom';

let checkbox;
let customCheckbox;
let withLabelCheckbox;

describe('custom checkbox', () => {
  beforeEach(() => {
    customCheckbox = (<Checkbox custom value={false} />);
  });

  it('should match snapshot', () => {
    const renderedCheckbox = renderer.create(customCheckbox);
    expect(renderedCheckbox.toJSON()).toMatchSnapshot();
  });

  it('should not be checked', () => {
    render(customCheckbox);
    const foundCheckbox = screen.getByRole('checkbox');
    expect(foundCheckbox).toBeInTheDocument();
    expect(foundCheckbox).not.toBeChecked();
  });

  it('should be checked after selecting', () => {
    const { rerender } = render(customCheckbox);
    const foundCheckbox = screen.getByRole('checkbox');
    expect(foundCheckbox).toBeInTheDocument();
    expect(foundCheckbox).not.toBeChecked();
    fireEvent.click(foundCheckbox);
    rerender(<Checkbox custom value />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).toBeChecked();
  });
});

describe('checkbox with label', () => {
  beforeEach(() => {
    withLabelCheckbox = (<Checkbox withLabel value={false} label="test label" />);
  });

  it('should match snapshot', () => {
    const renderedCheckbox = renderer.create(withLabelCheckbox);
    expect(renderedCheckbox.toJSON()).toMatchSnapshot();
  });

  it('should not be checked', () => {
    render(withLabelCheckbox);
    const foundCheckbox = screen.getByRole('checkbox');
    expect(foundCheckbox).toBeInTheDocument();
    expect(foundCheckbox).not.toBeChecked();
  });

  it('should be checked after selecting', () => {
    const { rerender } = render(withLabelCheckbox);
    const foundCheckbox = screen.getByRole('checkbox');
    expect(foundCheckbox).toBeInTheDocument();
    expect(foundCheckbox).not.toBeChecked();
    fireEvent.click(foundCheckbox);
    rerender(<Checkbox withLabel value label="test label" />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('should have label', () => {
    expect(screen.findByText('test label')).toBeTruthy();
  });
});

describe('default checkbox', () => {
  beforeEach(() => {
    checkbox = (<Checkbox value={false} />);
  });

  it('should match snapshot', () => {
    const renderedCheckbox = renderer.create(checkbox);
    expect(renderedCheckbox.toJSON()).toMatchSnapshot();
  });

  it('should not be checked', () => {
    render(checkbox);
    const foundCheckbox = screen.getByRole('checkbox');
    expect(foundCheckbox).toBeInTheDocument();
    expect(foundCheckbox).not.toBeChecked();
  });

  it('should be checked after selecting', () => {
    const { rerender } = render(checkbox);
    const foundCheckbox = screen.getByRole('checkbox');
    expect(foundCheckbox).toBeInTheDocument();
    expect(foundCheckbox).not.toBeChecked();
    fireEvent.click(foundCheckbox);
    rerender(<Checkbox custom value />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).toBeChecked();
  });
});
