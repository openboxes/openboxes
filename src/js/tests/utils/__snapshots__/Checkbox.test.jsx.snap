// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox with label should match snapshot 1`] = `
<div
  className="d-flex align-items-center"
  data-testid="with-label-checkbox"
>
  <input
    checked={false}
    id=""
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor=""
    style={
      {
        "margin": "0 0 0 5px",
      }
    }
  >
    test label
  </label>
</div>
`;

exports[`custom checkbox should match snapshot 1`] = `
<div
  className="custom-checkbox"
  data-testid="custom-checkbox"
>
  <label
    htmlFor=""
  >
    <input
      checked={false}
      id=""
      onChange={[Function]}
      type="checkbox"
    />
  </label>
</div>
`;

exports[`default checkbox should match snapshot 1`] = `
<input
  checked={false}
  data-testid="checkbox"
  id=""
  onChange={[Function]}
  type="checkbox"
/>
`;
